assayist/client/query.py

```diff
@@ -44,3 +46,114 @@ def get_container_content_sources(build_id):
+    * embeds artifacts that were built with artifacts built from the input and related source
+      locations
+    * was built with a container that embeds artifacts that were built with the input and related
+      source locations
+
+    :param list source_locations: a list of source location dictionaries to match against
+    :return: a list of affected container build Koji IDs
+    :rtype: list
+    """
+    if not source_locations or not isinstance(source_locations, collections.Iterable):
+        raise InvalidInput('The input must be a list of source locations')
+
+    # Get all the input source locations, then find the upstream or downstream source locations.
+    # With that result, find all the source locations that embed the resulting source locations.
+    # Then return the Neo4j IDs of the union of all the source locations in the query.
+    query = """
+    // First get all the input source locations
+    MATCH (input_sl:SourceLocation) WHERE input_sl.url IN [{0}]
+    // Then find all the source locations that are upstream or downstream of the input source
+    // locations recursively. The resulting `input_and_upstream_sl` variable has all the input
+    // source locations and all the upstream or downstream source locations of the input source
+    // locations.
+    MATCH (input_sl)-[:UPSTREAM*0..]-(input_and_upstream_sl:SourceLocation)
+    // Then find all the source locations that embed the source locations in
+    // `input_and_upstream_sl` recursively. The resulting `input_upstream_and_embedded_sl`
+    // variable will have the contents of the `input_and_upstream_sl` variable previously and
+    // all the source locations that eventually embed those source locations.
+    MATCH (input_and_upstream_sl)<-[:EMBEDS*0..]-(input_upstream_and_embedded_sl:SourceLocation)
+    RETURN ID(input_upstream_and_embedded_sl)
+    """.format(', '.join(repr(sl['url']) for sl in source_locations if 'url' in sl))
+    results, _ = neomodel.db.cypher_query(query)
+    # This should only be true if none of the input source locations are in the DB
+    if not results:
+        return []
+    all_sl_ids = set([r[0] for r in results])
+
+    # Find all the artifacts that were built from the source locations, and those that embed them.
+    # Then return the Neo4j IDs of the resulting artifacts.
+    query = """
+    // First get all the input source locations
+    MATCH (sl) WHERE ID(sl) IN [{0}]
+    // Find all the artifacts that were built from the source locations, and all the artifacts that
+    // embed those artifacts
+    MATCH (sl)<-[:BUILT_FROM]-(:Build)-[:PRODUCED]->(:Artifact)<-[:EMBEDS*0..]-(artifact:Artifact)
+    RETURN ID(artifact)
+    """.format(', '.join(str(sl_id) for sl_id in all_sl_ids))
+    results, _ = neomodel.db.cypher_query(query)
+    affected_artifact_ids = set([r[0] for r in results])
+
+    # Find all the builds of container artifacts that were built from any of the containers in
```

Comment:
Should this read "that were built with" for clarity?

Reply:
Sure, I can do that.